File: Pathway.cs

/*****************************************************************************
 *                                                                           *
 *  Pathway.cs                                                               *
 *  30 January 2007                                                          *
 *  Project: Crystalline                                                     *
 *  Written by: Richard Sartor                                               *
 *  Copyright © 2007 Metaphysics Industries, Inc.                            *
 *                                                                           *
 *  Converted from C++ to C# on 10 November 2007                             *
 *                                                                           *
 *  Acts as the pathway between two PathingJunction objects. Path objects    *
 *    follow it and turn at the junctions.                                   *
 *                                                                           *
 *****************************************************************************/

using System;
using System.Collections.Generic;
using MetaphysicsIndustries.Collections;
using System.Drawing;
using MetaphysicsIndustries.Utilities;

namespace MetaphysicsIndustries.Crystalline
{
    //public class Pathway : Box
    //{
    //    public Pathway()
    //    {
    //        this.Size = new SizeV(50, 25);
    //    }

    //    //public override void Dispose()
    //    //{
    //    //    LeftUp = null;
    //    //    RightDown = null;

    //    //    base.Dispose();
    //    //}

    //    public override void Render(Graphics g, Pen pen, Brush brush, Font font)
    //    {
    //        Graphics _g = g;
    //        RectangleV r = new RectangleV();
    //        if (pen == null)
    //        {
    //            pen = Pens.Black;
    //        }
    //        if (IsVertical)
    //        {
    //            if (LeftUp != null)
    //            {
    //                float x;
    //                SizeV size;
    //                if (RightDown != null)
    //                {
    //                    x = Math.Min(LeftUp.X, RightDown.X);
    //                    size = new SizeV(Math.Max(LeftUp.X, RightDown.X) + 25,
    //                    Math.Abs(LeftUp.Y - RightDown.Y) - 25);
    //                }
    //                else
    //                {
    //                    x = LeftUp.X;
    //                    size = new SizeV(25, 50);
    //                }

    //                r = new RectangleV(x, LeftUp.Y + 25, size.Width, size.Height);
    //            }
    //            else
    //            {
    //                if (RightDown != null)
    //                {
    //                    r = new RectangleV(RightDown.Location, new SizeV(25, 50));
    //                }
    //                else
    //                {
    //                    r = new RectangleV(0, 0, 0, 0);
    //                }
    //            }
    //        }
    //        else
    //        {
    //            if (LeftUp != null)
    //            {
    //                if (RightDown != null)
    //                {
    //                    r = new RectangleV(
    //                        Math.Min(LeftUp.X, RightDown.X) + 25,
    //                        Math.Min(LeftUp.Y, RightDown.Y),
    //                        Math.Abs(LeftUp.X - RightDown.X) - 25,
    //                        Math.Max(LeftUp.Y, RightDown.Y) + 25);
    //                }
    //                else
    //                {
    //                    r = new RectangleV(LeftUp.X + 25, LeftUp.Y, 50, 25);
    //                }
    //            }
    //            else
    //            {
    //                if (RightDown != null)
    //                {
    //                    r = new RectangleV(RightDown.X - 50, RightDown.Y, 50, 25);
    //                }
    //                else
    //                {
    //                    r = new RectangleV(0, 0, 0, 0);
    //                }
    //            }
    //        }


    //        r = this.Rect.Inflate( -1, -1);
    //        g.DrawRectangle(pen, r.X, r.Y, r.Width, r.Height);
    //    }

    //    public virtual bool IsVertical
    //    {
    //        get
    //        {
    //            return _isVertical;
    //        }
    //        set
    //        {
    //            if (_isVertical != value)
    //            {
    //                _isVertical = value;

    //                LeftUp = null;
    //                RightDown = null;


    //                //!@#$ --- fix this
    //                if (value)
    //                {
    //                    this.Width = 25;
    //                    this.Height = 50;
    //                }
    //                else
    //                {
    //                    this.Width = 50;
    //                    this.Height = 25;
    //                }
    //            }
    //        }
    //    }

    //    public virtual PathingJunction LeftUp
    //    {
    //        get
    //        {
    //            return _leftUp;
    //        }
    //        set
    //        {
    //            if (_leftUp != value)
    //            {
    //                if (_leftUp != null)
    //                {
    //                    if (IsVertical)
    //                    {
    //                        _leftUp.DownPathway = null;
    //                    }
    //                    else
    //                    {
    //                        _leftUp.RightPathway = null;
    //                    }
    //                }

    //                _leftUp = value;

    //                if (_leftUp != null)
    //                {
    //                    if (IsVertical)
    //                    {
    //                        _leftUp.DownPathway = this;
    //                    }
    //                    else
    //                    {
    //                        _leftUp.RightPathway = this;
    //                    }
    //                }

    //                //UpdateLeftUpRect();
    //            }
    //        }
    //    }

    //    public virtual PathingJunction RightDown
    //    {
    //        get
    //        {
    //            return _rightDown;
    //        }
    //        set
    //        {
    //            if (_rightDown != value)
    //            {
    //                if (_rightDown != null)
    //                {
    //                    if (IsVertical)
    //                    {
    //                        _rightDown.UpPathway = null;
    //                    }
    //                    else
    //                    {
    //                        _rightDown.LeftPathway = null;
    //                    }
    //                }

    //                _rightDown = value;

    //                if (_rightDown != null)
    //                {
    //                    if (IsVertical)
    //                    {
    //                        _rightDown.UpPathway = this;
    //                    }
    //                    else
    //                    {
    //                        _rightDown.LeftPathway = this;
    //                    }
    //                }

    //                //UpdateRightDownRect();
    //            }
    //        }
    //    }

    //    protected void UpdateLeftUpRect()
    //    {
    //        if (LeftUp == null) { return; }

    //        if (IsVertical)
    //        {
    //            float height = LeftUp.Height;
    //            LeftUp.Rect = new RectangleV(Left, Top - height, Width, height);
    //        }
    //        else
    //        {
    //            float width = LeftUp.Width;
    //            LeftUp.Rect = new RectangleV(Left - width, Top, width, Height);
    //        }
    //    }

    //    protected void UpdateRightDownRect()
    //    {
    //        if (RightDown == null) { return; }

    //        if (IsVertical)
    //        {
    //            float height = RightDown.Height;
    //            RightDown.Rect = new RectangleV(Left, Bottom, Width, height);
    //        }
    //        else
    //        {
    //            float width = RightDown.Width;
    //            RightDown.Rect = new RectangleV(Right, Top, width, Height);
    //        }
    //    }

    //    protected void UpdateLeftUpRectByRelocate()
    //    {
    //        if (LeftUp == null) { return; }

    //        RectangleV rect;

    //        if (IsVertical)
    //        {
    //            float height = LeftUp.Height;
    //            rect = new RectangleV(Left, Top - height, Width, height);
    //        }
    //        else
    //        {
    //            float width = LeftUp.Width;
    //            rect = new RectangleV(Left - width, Top, width, Height);
    //        }

    //        //change this to Framework.Relocate
    //        LeftUp.Size = rect.Size;
    //        Framework.Move(LeftUp, rect.Location);
    //    }

    //    protected void UpdateRightDownRectByRelocate()
    //    {
    //        if (RightDown == null) { return; }

    //        RectangleV rect;

    //        if (IsVertical)
    //        {
    //            float height = RightDown.Height;
    //            rect = new RectangleV(Left, Bottom, Width, height);
    //        }
    //        else
    //        {
    //            float width = RightDown.Width;
    //            rect = new RectangleV(Right, Top, width, Height);
    //        }

    //        //change this to Framework.Relocate
    //        RightDown.Size = rect.Size;
    //        Framework.Move(RightDown, rect.Location);
    //    }

    //    protected override void OnRectChanged(EventArgs e)
    //    {
    //        //UpdateLeftUpRectByRelocate();
    //        //UpdateRightDownRectByRelocate();

    //        base.OnRectChanged(e);
    //    }

    //    private bool _isVertical;
    //    private PathingJunction _leftUp;
    //    private PathingJunction _rightDown;
    //}
}

File: Crystalline/Pathway.cpp

/*****************************************************************************
 *                                                                           *
 *  Pathway.cpp                                                              *
 *  30 January 2007                                                          *
 *  Project: Crystalline                                                     *
 *  Written by: Richard Sartor                                               *
 *  Copyright © 2007 Metaphysics Industries                                  *
 *                                                                           *
 *  Acts as the pathway between two PathingJunction objects. Path objects    *
 *    follow it and turn at the junctions.                                   *
 *                                                                           *
 *****************************************************************************/



#include "stdafx.h"
#include "Pathway.h"



;
NAMESPACE_START
;



Pathway::Pathway(void)
{
	this->Size = SizeF(50, 25);
}

Pathway::~Pathway(void)
{
}

void Pathway::Render(Graphics^ g, Pen^ pen)
{
	Graphics^ _g = g;

	RectangleF	r;

	if (pen == nullptr)
	{
		pen = Pens::Black;
	}

	if (this->IsVertical)
	{
		if (this->LeftUp)
		{
			if (this->RightDown)
			{
				r.X = Math::Min(this->LeftUp->X, this->RightDown->X);
				r.Width = Math::Max(this->LeftUp->X, this->RightDown->X) + 25;
				r.Height = Math::Abs(this->LeftUp->Y - this->RightDown->Y) - 25;
			}
			else
			{
				r.X = this->LeftUp->X;
				r.Width = 25;//this->LeftUp->Width;
				r.Height = 50;
			}
			r.Y = this->LeftUp->Y + 25;
		}
		else
		{
			if (this->RightDown)
			{
				r.X = this->RightDown->X;
				r.Width = 25;
				r.Y = this->RightDown->Y - 50;
				r.Height = 50;
			}
			else
			{
				r = RectangleF(0, 0, 0, 0);
			}
		}
	}
	else
	{
		if (this->LeftUp)
		{
			if (this->RightDown)
			{
				r.X = Math::Min(this->LeftUp->X, this->RightDown->X) + 25;
				r.Y = Math::Min(this->LeftUp->Y, this->RightDown->Y);
				r.Width = Math::Abs(this->LeftUp->X - this->RightDown->X) - 25;
				r.Height = Math::Max(this->LeftUp->Y, this->RightDown->Y) + 25;
			}
			else
			{
				r.X = this->LeftUp->X + 25;
				r.Y = this->LeftUp->Y;
				r.Width = 50;
				r.Height = 25;
			}
		}
		else
		{
			if (this->RightDown)
			{
				r.X = this->RightDown->X - 50;
				r.Y = this->RightDown->Y;
				r.Width = 50;
				r.Height = 25;
			}
			else
			{
				r = RectangleF(0, 0, 0, 0);
			}
		}
	}

	//r = RectangleF(this->X, this->Y, 25, 25);

	//g->DrawRectangle(pen, r.X+1, r.Y+1, r.Width-2, r.Height-2);
	r = RectangleF::Inflate(this->Rect, -1, -1);
	g->DrawRectangle(pen, r.X, r.Y, r.Width, r.Height);
}

PathingJunction^ Pathway::LeftUp::get(void)
{
	return this->leftup;
}
void Pathway::LeftUp::set(PathingJunction^ pj)
{
	if (this->leftup != pj)
	{
		this->leftup = pj;
	}
}

PathingJunction^ Pathway::RightDown::get(void)
{
	return this->rightdown;
}
void Pathway::RightDown::set(PathingJunction^ pj)
{
	if (this->rightdown != pj)
	{
		this->rightdown = pj;
	}
}

Element^ Pathway::LeftUpElement::get(void)
{
	return this->leftupelement;
}
void Pathway::LeftUpElement::set(Element^ e)
{
	if (this->leftupelement != e)
	{
		this->leftupelement = e;

		if (this->LeftUp && this->RightDown)
		{
			if (this->IsVertical)
			{
				this->LeftUp->X = this->RightDown->X;
			}
			else
			{
				this->LeftUp->Y = this->RightDown->Y;
			}
		}
		if (this->IsVertical)
		{
			this->LeftUp->DownPathway = this;
		}
		else
		{
			this->RightDown->UpPathway = this;
		}
	}
}

Element^ Pathway::RightDownElement::get(void)
{
	return this->rightdownelement;
}
void Pathway::RightDownElement::set(Element^ e)
{
	if (this->rightdownelement != e)
	{
		this->rightdownelement = e;
	}
}

bool Pathway::IsVertical::get(void)
{
	return this->isvertical;
}
void Pathway::IsVertical::set(bool b)
{
	if (this->isvertical != b)
	{
		this->isvertical = b;
		if (b)
		{
			this->Width = 25;
			this->Height = 50;
		}
		else
		{
			this->Width = 50;
			this->Height = 25;
		}
	}
}






;
NAMESPACE_END
;

